src/linker/Linker.Dataflow/MethodBodyScanner.cs

```diff
@@ -173,17 +174,19 @@ public int MoveNext (Instruction op)
 			}
 		}
 
-		static void ValidateLocalStore (LocalVariableStore locals)
+		[Conditional ("DEBUG")]
+		static void ValidateLocalStore (LocalVariableStore locals, MethodDefinition method, int ilOffset)
 		{
-			foreach (var value in locals.Keys) {
-				foreach (var val in locals[value].Value) {
+			foreach (var variableDef in locals.Keys) {
+				foreach (var val in locals[variableDef].Value) {
 					if (val is LocalVariableReferenceValue reference
 						&& locals[reference.LocalDefinition].Value.Any (v => v is ReferenceValue)) {
-						throw new Exception ();
+						throw new LinkerFatalErrorException (MessageContainer.CreateCustomErrorMessage ($"Linker error: Local variable {variableDef.Index}  references variable {reference.LocalDefinition.Index} which is a reference.", 6002, origin: new MessageOrigin (method, ilOffset)));
```

Comment:
The error code 6002 is really weird. I think it should be something in the 1xxx range. In other places we add it to DiagnosticId and also use resources I think.